{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [-h] [-I Input_file] [-O Output_file]\n",
      "                             [-l peptide_length]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: -f C:\\Users\\DanielNZG85\\AppData\\Roaming\\jupyter\\runtime\\kernel-5c90e740-9aa8-436b-8948-fb001784cf82.json\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[1;31mSystemExit\u001b[0m\u001b[1;31m:\u001b[0m 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\DanielNZG85\\Anaconda2\\envs\\py36\\lib\\site-packages\\IPython\\core\\interactiveshell.py:2969: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "#neoExtract -  sorts the amino-acid sequences etracted from annovar gene annotation and only contains the mutated amino-acid of specify length\n",
    "#Typically Neoantigens are 8-11 ammino acid long although can often reach up to 14-mer\n",
    "#This script inputs a fasta file containing the coding sequences and outputs another fasta file with the sequences \n",
    "#WILDTYPE proteins are removed\n",
    "#Lines starting with '>' are mantain as the header of the protein sequence\n",
    "\n",
    "from pylab import *\n",
    "import re\n",
    "import numpy as np\n",
    "import sys\n",
    "import argparse\n",
    "\n",
    "\n",
    "input_file = 'filename'\n",
    "output_file = 'outputfile'\n",
    "length_peptide = 'length'\n",
    "\n",
    "filename=open(input_file,'r');\n",
    "m=filename.read();\n",
    "filename.close();\n",
    "#input_invalid = 1\n",
    "#while input_invalid:\n",
    "#    length_peptide = int(float(input(\"Please enter maximum amino acid length of the neoantigens (from 8 to 14 mer): \")))\n",
    "#    if length_peptide <= 14:\n",
    "#        input_invalid = 0\n",
    "#        length_peptide = length_peptide-1\n",
    "#    else:\n",
    "#        print(\"Please enter a valid integer\")\n",
    "\n",
    "\n",
    "\n",
    "#Divide into headers\n",
    "d=m.replace('\\n','').replace('>','@>');\n",
    "D={};\n",
    "D=d.split('@');\n",
    "del D[0]\n",
    "\n",
    "#Remove WildType Sequences\n",
    "D1 = [];\n",
    "for k in range(0,len(D)):\n",
    "    if not re.search('WILDTYPE', D[k]):\n",
    "        D1.append(D[k])\n",
    "    \n",
    "#Separete sequence from header    \n",
    "C3=[];\n",
    "for i in range (0, len(D1)):\n",
    "    p = re.split('\\)', D1[i])\n",
    "    p[0] = p[0]+')'\n",
    "    C3.append(p[0])\n",
    "    C3.append(p[1])\n",
    "\n",
    "\n",
    "index_mut = [];\n",
    "#Separate depending on the type of mutation and extract indexes of mutation\n",
    "for i in range (0, len(C3)):\n",
    "    if re.search('insertion', C3[i]):\n",
    "        result = re.search('position (.*) has', C3[i])\n",
    "        index_mut.append(result.group(1))\n",
    "    elif re.search('changed', C3[i]):\n",
    "        result = re.search('position (.*) changed', C3[i])\n",
    "        index_mut.append(result.group(1))\n",
    "    elif re.search('amino', C3[i]):\n",
    "        result = re.search('position (.*) changed', C3[i])\n",
    "        index_mut.append('0')\n",
    "        \n",
    "\n",
    "#Separate initial and end indexes and create master matrix\n",
    "master_matrix=[];\n",
    "p=0;\n",
    "rt=0;\n",
    "for i in range (0,len(index_mut)):\n",
    "    if re.search('-',index_mut[i]):    \n",
    "        h1 = re.split('-', index_mut[i])\n",
    "    else:\n",
    "        rt = 1+rt;\n",
    "        h1 = [index_mut[i],'0'];\n",
    "    master_matrix.append([h1[0], h1[1], C3[p],C3[p+1]]);\n",
    "    p=p+2;\n",
    "    \n",
    "    \n",
    "#Sort peptide sequences accordingly\n",
    "just_seq =[];\n",
    "final_seq=[];\n",
    "\n",
    "for i in range (0, len(master_matrix)):\n",
    "    if len(master_matrix[i][2]) >= 2000:\n",
    "        line_index = master_matrix[i][2][0:2000];\n",
    "    else:\n",
    "        line_index = master_matrix[i][2];\n",
    "        \n",
    "    neat_seq = master_matrix[i][3];\n",
    "    if re.search('to',master_matrix[i][2]):\n",
    "        prem_seq=re.search(' to (.*)\\)', master_matrix[i][2])\n",
    "        if prem_seq:\n",
    "            exact_seq = prem_seq.group(1)\n",
    "        else:\n",
    "            exact_seq = '';\n",
    "    else:\n",
    "        prem_seq=re.search('insertion (.*)\\)', master_matrix[i][2])\n",
    "        if prem_seq:\n",
    "            exact_seq = prem_seq.group(1)\n",
    "        else:\n",
    "            exact_seq='';\n",
    "\n",
    "    if int(float(master_matrix[i][0])) == 0:\n",
    "        cut_seq = master_matrix[i][3];\n",
    "    elif int(float(master_matrix[i][1])) == 0:     \n",
    "        cut_seq = master_matrix[i][3];\n",
    "        if re.search('immediate-stoploss',master_matrix[i][2]):\n",
    "            cut_seq = neat_seq[len(neat_seq)-length_peptide-len(exact_seq):len(neat_seq)]\n",
    "        elif float(master_matrix[i][0]) > len(neat_seq):\n",
    "            cut_seq = neat_seq;\n",
    "        elif (float(master_matrix[i][0])-length_peptide) < 1 and (float(master_matrix[i][0])+length_peptide+len(exact_seq)) > len(neat_seq):\n",
    "            cut_seq = neat_seq[0:len(neat_seq)]\n",
    "        elif float(master_matrix[i][0])-length_peptide < 1:\n",
    "            cut_seq = neat_seq[0:int(float(master_matrix[i][0]))+length_peptide+len(exact_seq)]\n",
    "        elif float(master_matrix[i][0])+length_peptide+len(exact_seq) > len(neat_seq):\n",
    "            cut_seq = neat_seq[int(float(master_matrix[i][0]))-length_peptide:len(neat_seq)]\n",
    "        else:\n",
    "            cut_seq = neat_seq[int(float(master_matrix[i][0]))-length_peptide+1:int(float(master_matrix[i][0]))+length_peptide+len(exact_seq)]\n",
    "    else:\n",
    "        if len(neat_seq)==len(exact_seq):\n",
    "            cut_seq = neat_seq;\n",
    "        elif float(master_matrix[i][0])-length_peptide-1 < 1 and  float(master_matrix[i][1])+length_peptide > len(neat_seq):\n",
    "            cut_seq = neat_seq[0:len(neat_seq)]\n",
    "        elif float(master_matrix[i][0])-length_peptide-1 < 1:\n",
    "            cut_seq = neat_seq[0:int(float(master_matrix[i][1]))+length_peptide]\n",
    "        elif float(master_matrix[i][1])+length_peptide > len(neat_seq):\n",
    "            cut_seq = neat_seq[int(float(master_matrix[i][0]))-length_peptide:len(neat_seq)]\n",
    "        elif float(master_matrix[i][1])+length_peptide+len(exact_seq) > len(neat_seq):\n",
    "            cut_seq = neat_seq[int(float(master_matrix[i][0]))-length_peptide:int(float(master_matrix[i][0]))-1+len(exact_seq)]\n",
    "        else:\n",
    "            cut_seq = neat_seq[int(float(master_matrix[i][0]))-length_peptide:int(float(master_matrix[i][0]))+len(exact_seq)+length_peptide]\n",
    "    \n",
    "    final_seq.append(line_index)\n",
    "    final_seq.append(cut_seq)\n",
    "    just_seq.append(cut_seq)\n",
    "    \n",
    "    \n",
    "#Calculate average length of the peptides just for validation\n",
    "length_index = [];\n",
    "for i in range (1,len(just_seq)):\n",
    "    length_index.append(len(just_seq[i]))\n",
    "ave_length= sum(length_index)/len(length_index)\n",
    "\n",
    "#Write output to file\n",
    "with open(output_file, 'w') as f:\n",
    "    for item in final_seq:\n",
    "        f.write(\"%s\\n\" % item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
